File: 5-tamper-evident-audit-logging/1-java/src/main/java/com/starci/auditapi/AuditController.java
package com.starci.auditapi;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

import java.util.Map;

@RestController
public class AuditController {

    private final AuditService auditService;

    public AuditController(AuditService auditService) {
        this.auditService = auditService;
    }

    // POST /events — append one event, returning index + the two hashes (HTTP 201).
    @PostMapping("/events")
    public ResponseEntity<Map<String, Object>> append(@RequestBody(required = false) Map<String, String> body) {
        Map<String, String> b = body == null ? Map.of() : body;
        String actor = b.getOrDefault("actor", "unknown");
        String action = b.getOrDefault("action", "unknown");
        String target = b.getOrDefault("target", "unknown");
        return ResponseEntity.status(HttpStatus.CREATED).body(auditService.append(actor, action, target));
    }

    // GET /verify — recompute the chain and report OK or the first broken index.
    @GetMapping("/verify")
    public Map<String, Object> verify() {
        return auditService.verify();
    }
}

File: 5-tamper-evident-audit-logging/0-typescript/src/audit/audit.controller.ts
import { Body, Controller, Get, HttpCode, Post } from "@nestjs/common"
import { AuditService } from "./audit.service"

interface AppendDto {
    actor?: string
    action?: string
    target?: string
}

@Controller()
export class AuditController {
    constructor(private readonly auditService: AuditService) {}

    // POST /events — append one event, returning its position and the two hashes
    // so the caller can independently verify linkage.
    @Post("events")
    @HttpCode(201)
    async append(@Body() body: AppendDto) {
        const actor = body.actor ?? "unknown"
        const action = body.action ?? "unknown"
        const target = body.target ?? "unknown"
        return this.auditService.append(actor, action, target)
    }

    // GET /verify — recompute the entire chain and report OK or the first broken index.
    @Get("verify")
    async verify() {
        return this.auditService.verify()
    }
}
